raytracing/drawing.py

```diff
@@ -0,0 +1,324 @@
+
+        >>> fig, axes = plt.subplots()
+        >>> lensDrawing.applyTo(axes, x=10)
+
+        Take advantage of the auto-scaling feature by updating the Drawing after the figure's limits have changed
+        (on a zoom callback)
+
+        >>> lensDrawing.update()
+
+        Update the Drawing's position
+
+        >>> lensDrawing.update(x=5)
+
+    """
+
+    def __init__(self, components: List[patches.Patch], label: str = None,
+                 x=0, y=0, fixedWidth=False):
+        self.components = components
+        self.label = None
+
+        self.axes = None
+        self.x = x
+        self.y = y
+        self.useAutoScale = not fixedWidth
+
+        if label is not None:
+            self.label = Label(text=label, x=self.centroid[0], y=self.halfHeight() * 1.2)
+
+    @property
+    def hasLabel(self) -> bool:
+        if self.label is None:
+            return False
+        return True
+
+    def applyTo(self, axes: plt.Axes, x: float = None, y: float = None):
+        """ Apply the Drawing on a figure at a given position (x, y) with auto-scale.
+
+        Args:
+            axes (matplotlib.pyplot.Axes): The figure's Axes on which to apply the drawing.
+            x, y (:obj:`float`, optional): The x and y position in data units where to apply the drawing.
+                Defaults to (0, 0).
+
+        """
+
+        self.axes = axes
+        if x is not None:
+            self.x = x
+        if y is not None:
+            self.y = y
+
+        self.update()
+
+        for component in self.components:
+            self.axes.add_patch(component)
+
+        if self.label is not None:
+            self.axes.add_artist(self.label)
+
+    def update(self, x: float = None, y: float = None):
+        """ Update the drawing's position and scaling.
+
+        Args:
+            x, y (:obj:`float`, optional): The x and y position where to apply the drawing.
+                Defaults to the originally applied position.
+
+        """
+
+        if x is not None:
+            self.x = x
+        if y is not None:
+            self.y = y
+
+        xScaling, yScaling = self.scaling()
+
+        translation = transforms.Affine2D().translate(self.x, self.y)
+        scaling = transforms.Affine2D().scale(xScaling, yScaling)
+
+        for component in self.components:
+            component.set_transform(scaling + translation + self.axes.transData)
+
+        if self.label is not None:
+            self.label.set_transform(translation + self.axes.transData)
+
+    def scaling(self):
+        """ Used internally to compute the required scale transform so that the width of the objects stay the same
+        respective to the Axes. """
+        if not self.useAutoScale:
+            return 1, 1
+
+        xScale, yScale = self.axesToDataScale()
+
+        heightFactor = self.halfHeight() * 2 / yScale
+        xScaling = xScale * (heightFactor / 0.2) ** (3 / 4)
+
+        return xScaling, 1
+
+    def axesToDataScale(self):
```

Comment:
This would be overridden in a `MatplotlibGraphics`, `CocoaGraphics`, `TkGraphics`, etc... Everything else does not need to know the details of the backend.